### PR TITLE
Add BuildConfig object

### PR DESCRIPTION
**Pull request by sgallagher**
Create the basic object type and getters/setters.

Create the YAML parser.

Create unit tests for BuildConfig.

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

### CODE REVIEW

**Comment by sgallagher** [comment]
I think the `ModulemdBuildConfig` object is ready to be reviewed now. I implemented the object, the setters/getters and the YAML parser (as well as unit tests for all of those).

It does not include a YAML emitter at this time, because I decided that was "nice-to-have" for debugging, but not important enough to spend time on while under time-pressure.

The `ModulemdPackagerV3` object should be in an equivalent state tomorrow.

File: modulemd/tests/test-modulemd-build-config.c [hunk 1]
```diff
@@ -0,0 +1,733 @@
...
+buildconfig_test_parse_yaml_no_context (void)
+{
+  MMD_INIT_YAML_EVENT (event);
+  MMD_INIT_YAML_PARSER (parser);
+  g_autofree gchar *yaml_path = NULL;
+  g_autoptr (ModulemdBuildConfig) bc = NULL;
+  g_autoptr (FILE) yaml_stream = NULL;
+  g_autoptr (GError) error = NULL;
+
+  /* Read a good YAML file with an unknown key */
```
Comment (mmathesius) [inline modulemd/tests/test-modulemd-build-config.c:554]:
```suggestion
  /* Read a YAML file that is missing context */
```

File: modulemd/tests/test-modulemd-build-config.c [hunk 2]
```diff
@@ -0,0 +1,733 @@
...
+buildconfig_test_parse_yaml_short_context (void)
+{
+  MMD_INIT_YAML_EVENT (event);
+  MMD_INIT_YAML_PARSER (parser);
+  g_autofree gchar *yaml_path = NULL;
+  g_autoptr (ModulemdBuildConfig) bc = NULL;
+  g_autoptr (FILE) yaml_stream = NULL;
+  g_autoptr (GError) error = NULL;
+
+  /* Read a good YAML file with an unknown key */
```
Comment (mmathesius) [inline modulemd/tests/test-modulemd-build-config.c:582]:
```suggestion
  /* Read a YAML file with an empty context */
```

File: modulemd/tests/test-modulemd-build-config.c [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+buildconfig_test_parse_yaml_long_context (void)
+{
+  MMD_INIT_YAML_EVENT (event);
+  MMD_INIT_YAML_PARSER (parser);
+  g_autofree gchar *yaml_path = NULL;
+  g_autoptr (ModulemdBuildConfig) bc = NULL;
+  g_autoptr (FILE) yaml_stream = NULL;
+  g_autoptr (GError) error = NULL;
+
+  /* Read a good YAML file with an unknown key */
```
Comment (mmathesius) [inline modulemd/tests/test-modulemd-build-config.c:610]:
```suggestion
  /* Read a YAML file that has a context that is too long */
```

File: modulemd/tests/test-modulemd-build-config.c [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+buildconfig_test_parse_yaml_nonalphanum (void)
+{
+  MMD_INIT_YAML_EVENT (event);
+  MMD_INIT_YAML_PARSER (parser);
+  g_autofree gchar *yaml_path = NULL;
+  g_autoptr (ModulemdBuildConfig) bc = NULL;
+  g_autoptr (FILE) yaml_stream = NULL;
+  g_autoptr (GError) error = NULL;
+
+  /* Read a good YAML file with an unknown key */
```
Comment (mmathesius) [inline modulemd/tests/test-modulemd-build-config.c:638]:
```suggestion
  /* Read a YAML file that has a context with a disallowed character */
```

File: modulemd/tests/test-modulemd-build-config.c [hunk 5]
```diff
@@ -0,0 +1,733 @@
...
+buildconfig_test_parse_yaml_no_platform (void)
+{
+  MMD_INIT_YAML_EVENT (event);
+  MMD_INIT_YAML_PARSER (parser);
+  g_autofree gchar *yaml_path = NULL;
+  g_autoptr (ModulemdBuildConfig) bc = NULL;
+  g_autoptr (FILE) yaml_stream = NULL;
+  g_autoptr (GError) error = NULL;
+
+  /* Read a good YAML file with an unknown key */
```
Comment (mmathesius) [inline modulemd/tests/test-modulemd-build-config.c:666]:
```suggestion
  /* Read a YAML file that is missing platform */
```

**Review by mmathesius — CHANGES_REQUESTED** [review]

File: modulemd/include/private/modulemd-build-config.h [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+ * configuration.
+ *
+ * Since: 2.10
+ */
+ModulemdBuildConfig *
+modulemd_build_config_new (void);
+
+
+/**
+ * MMD_MAXCONTEXTLEN:
```
[the remark this answers is not in the document]
Comment (sgallagher) [reply modulemd/include/private/modulemd-build-config.h:48]: Yeah, probably it should be.

File: modulemd/include/private/modulemd-build-config.h [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+#pragma once
+
+#include <glib-object.h>
+#include "modulemd-2.0/modulemd-buildopts.h"
+#include "private/modulemd-yaml.h"
+
+G_BEGIN_DECLS
+
+/**
+ * SECTION: modulemd-build-config
```
[the remark this answers is not in the document]
Comment (sgallagher) [reply modulemd/include/private/modulemd-build-config.h:23]: I intentionally didn't name it `-private` which is a thing we only do for objects that have both a public and a private header.

File: modulemd/modulemd-build-config.c [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+        {
+          g_set_error (error,
+                       MODULEMD_ERROR,
+                       MMD_ERROR_VALIDATE,
+                       "Non-alphanumeric character in BuildConfig context");
+          return FALSE;
+        }
+    }
+
+  if (i >= MMD_MAXCONTEXTLEN)
```
[the remark this answers is not in the document]
Comment (sgallagher) [reply modulemd/modulemd-build-config.c:434]: "There are only two hard problems in computer science..."

**Comment by sgallagher** [comment]
I fixed up the comments except for renaming the header to `-private`.

**Pull request merged**